LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/Regions/Region.cs

```diff
@@ -157,52 +157,69 @@ public bool TryGetUpstreamChannelFrequency(Rxpk upstreamChannel, out double freq
+        {
+            // If the rx2 datarate property is in twins, we take it from there
+            if (rx2DrFromTwins.HasValue)
             {
-                // Otherwise we check if we have some properties set on the server (server Specific)
-                if (string.IsNullOrEmpty(nwkSrvRx2Dr))
+                if (this.RegionLimits.IsCurrentDownstreamDRIndexWithinAcceptableValue(rx2DrFromTwins))
                 {
-                    // If not we use the region default.
-                    Logger.Log(devEUI, $"using standard second receive windows for join request", LogLevel.Debug);
-                    // using EU fix DR for RX2
-                    freq = this.RX2DefaultReceiveWindows.frequency;
-                    datr = this.DRtoConfiguration[this.RX2DefaultReceiveWindows.dr].configuration;
+                    var datr = this.DRtoConfiguration[rx2DrFromTwins.Value].configuration;
+                    Logger.Log(devEUI, $"using device twins rx2: {rx2DrFromTwins.Value}, datr: {datr}", LogLevel.Debug);
+                    return datr;
                 }
                 else
                 {
-                    Logger.Log(devEUI, $"using custom second receive windows for join request", LogLevel.Debug);
-                    freq = nwkSrvRx2Freq;
-                    datr = nwkSrvRx2Dr;
+                    var datr = this.DRtoConfiguration[this.RX2DefaultReceiveWindows.dr].configuration;
+                    Logger.Log(devEUI, $"device twins rx2 ({rx2DrFromTwins.Value}) is invalid, using default: {this.RX2DefaultReceiveWindows.dr}, datr: {datr}", LogLevel.Debug);
+                    return datr;
                 }
             }
             else
             {
-                uint rx2Dr = (uint)rx2DrFromTwins;
-                if (this.RegionLimits.IsCurrentDownstreamDRIndexWithinAcceptableValue(rx2Dr))
+                // Otherwise we check if we have some properties set on the server (server Specific)
+                if (string.IsNullOrEmpty(nwkSrvRx2Dr))
                 {
-                    datr = this.DRtoConfiguration[rx2Dr].configuration;
-                    Logger.Log(devEUI, $"using device rx2: {rx2Dr}, datr: {datr}, region: {this.LoRaRegion}", LogLevel.Debug);
+                    // If not we use the region default.
+                    var datr = this.DRtoConfiguration[this.RX2DefaultReceiveWindows.dr].configuration;
+                    Logger.Log(devEUI, $"using standard region RX2 datarate {datr}", LogLevel.Debug);
+                    return datr;
                 }
                 else
                 {
-                    datr = this.DRtoConfiguration[this.RX2DefaultReceiveWindows.dr].configuration;
-                    Logger.Log(devEUI, $"device rx2 ({rx2Dr}) is invalid, using default: {this.RX2DefaultReceiveWindows.dr}, datr: {datr}, region: {this.LoRaRegion}", LogLevel.Debug);
+                    var datr = nwkSrvRx2Dr;

```

Comment:
can we refactor the 2 default cases and put them at the end. we return early if we have other configurations.